src/runtime/test/render-vdom.spec.tsx

```diff
@@ -1097,5 +1097,93 @@ describe('render-vdom', () => {
       await waitForChanges();
       expect(rootInstance.counter).toEqual(2);
     });
+
+    it('should not call ref cb w/ null when children are reordered', async () => {
+      @Component({ tag: 'cmp-a' })
+      class CmpA {
+        counter = 0;
+        divRef: HTMLElement;
+        @Prop() state = true;
+
+        renderA() {
+          return (
+            <div class="a" ref={(el) => (this.divRef = el)}>
+              A
+            </div>
+          );
+        }
+
+        renderB() {
+          return <div>B</div>;
+        }
+
+        render() {
+          return this.state
+            ? [this.renderB(), <div>middle</div>, this.renderA()]
+            : [this.renderA(), <div>middle</div>, this.renderB()];
+        }
+      }
+
+      const { root, rootInstance, waitForChanges } = await newSpecPage({
+        components: [CmpA],
+        html: `<cmp-a></cmp-a>`,
+      });
+
+      root.state = false;
+      await waitForChanges();
+      // We've changed the state and forced a re-render. This tests one of the
+      // ways in which children can be re-ordered that the `updateChildren` algo
+      // can handle without having `key` attrs set.
+      expect(rootInstance.divRef).toEqual(root.querySelector('.a'));
+    });
+
+    it('should not call ref cb w/ null when children w/ keys are reordered', async () => {
+      @Component({ tag: 'cmp-a' })
+      class CmpA {
+        counter = 0;
+        divRef: HTMLElement;
+        @Prop() state = true;
+
+        renderA() {
+          return (
+            <div key="a" class="a" ref={(el) => (this.divRef = el)}>
+              A
+            </div>
+          );
+        }
+
+        renderB() {
+          return <div>B</div>;
+        }
+
+        render() {
+          return this.state
+            ? [this.renderB(), this.renderA(), this.renderB(), this.renderB()]
+            : [this.renderA(), this.renderB(), this.renderB()];
+        }
+      }
+
+      const { root, rootInstance, waitForChanges } = await newSpecPage({
+        components: [CmpA],
+        html: `<cmp-a></cmp-a>`,
+      });
+
```

Comment:
Does it make sense for us to put an assertion before we force the re-render that `divRef` is `undefined`?
```suggestion
expect(rootInstance.divRef).toBeUndefined();
```

That way, we know that the re-render definitely happened, as opposed to a scenario where something funky happens and the contents of `renderA()` were being rendered the whole time